keadm/cmd/keadm/app/cmd/edge/rollback.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2022 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package edge
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/yaml"
+
+	"github.com/kubeedge/kubeedge/common/constants"
+	"github.com/kubeedge/kubeedge/edge/pkg/common/util"
+	"github.com/kubeedge/kubeedge/pkg/apis/componentconfig/edgecore/v1alpha2"
+	api "github.com/kubeedge/kubeedge/pkg/apis/fsm/v1alpha1"
+	"github.com/kubeedge/kubeedge/pkg/util/fsm"
+	"github.com/kubeedge/kubeedge/pkg/version"
+)
+
+// NewEdgeUpgrade returns KubeEdge edge upgrade command.
+func NewEdgeRollback() *cobra.Command {
+	rollbackOptions := newRollbackOptions()
+
+	cmd := &cobra.Command{
+		Use:   "rollback",
+		Short: "rollback edge component. Rollback the edge node to the desired version.",
+		Long:  "Rollback edge component. Rollback the edge node to the desired version.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// upgrade edgecore
+			return rollbackOptions.rollback()
+		},
+	}
+
+	addRollbackFlags(cmd, rollbackOptions)
+	return cmd
+}
+
+// newJoinOptions returns a struct ready for being used for creating cmd join flags.
+func newRollbackOptions() *RollbackOptions {
+	opts := &RollbackOptions{}
+	opts.HistoryVersion = version.Get().String()
+	opts.Config = constants.DefaultConfigDir + "edgecore.yaml"
+
+	return opts
+}
+
+func (ro *RollbackOptions) rollback() error {
```

Comment:
The `rollback(..)` function may not be suitable as a `RollbackOptions{}` struct method. It's better to define it as a function: `func rollback(opts RollbackOptions) error {}`

Reply:
I don't understand. Can you explain it?

Reply:
> I don't understand. Can you explain it?

By design, options are provided for the rollback(), the rollback() is not a method to options, so it's better to use the function.
```golang
type RollbackOptions { ... }

func rollback(opts *RollbackOptions) error { ... }
```